hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/NeighborRankTraverser.java

```diff
@@ -0,0 +1,270 @@
+                                                        step.capacity);
+                this.contributeLastLayer(sameLayerIncrRanks, lastLayerRanks);
+            }
+            ranks.add(newLayerRanks);
+
+            // Re-init sources
+            sources = newVertices;
+        }
+        return this.topRanks(ranks, steps);
+    }
+
+    private boolean belongToSameLayer(Set<Id> sources, Id target,
+                                      Set<Id> sameLayerNodes) {
+        if (sources.contains(target)) {
+            sameLayerNodes.add(target);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean belongToPrevLayers(List<Ranks> ranks, Id target,
+                                       Map<Integer, Set<Id>> prevLayerNodes) {
+        for (int i = ranks.size() - 2; i > 0; i--) {
+            Ranks prevLayerRanks = ranks.get(i);
+            if (prevLayerRanks.containsKey(target)) {
+                Set<Id> nodes = prevLayerNodes.computeIfAbsent(i, HashSet::new);
+                nodes.add(target);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void mergeSameLayerIncrRanks(Set<Id> sameLayerNodesV, double incr,
+                                         Map<Id, Double> sameLayerIncrRanks) {
+        for (Id node : sameLayerNodesV) {
+            double oldRank = sameLayerIncrRanks.getOrDefault(node, 0.0);
+            sameLayerIncrRanks.put(node, oldRank + incr);
+        }
+    }
+
+    private void contributePrevLayers(List<Ranks> ranks, double incr,
+                                      Map<Integer, Set<Id>> prevLayerNodesV) {
+        for (Map.Entry<Integer, Set<Id>> e : prevLayerNodesV.entrySet()) {
+            Ranks prevLayerRanks = ranks.get(e.getKey());
+            for (Id node : e.getValue()) {
+                double oldRank = prevLayerRanks.get(node);
+                prevLayerRanks.put(node, oldRank + incr);
+            }
+        }
+    }
+
+    private void contributeLastLayer(Map<Id, Double> rankIncrs,
+                                     Ranks lastLayerRanks) {
+        for (Map.Entry<Id, Double> entry : rankIncrs.entrySet()) {
+            double originRank = lastLayerRanks.get(entry.getKey());
+            double incrRank = entry.getValue();
+            lastLayerRanks.put(entry.getKey(), originRank + incrRank);
+        }
+    }
+
+    private Ranks contributeNewLayer(MultivaluedMap<Id, Node> adjacencies,
+                                     Ranks lastLayerRanks, int capacity) {
+        Ranks newLayerRanks = new Ranks(capacity);
+        for (Map.Entry<Id, List<Node>> entry : adjacencies.entrySet()) {
+            Id parent = entry.getKey();
+            long size = entry.getValue().size();
+            for (Node node : entry.getValue()) {
+                double rank = newLayerRanks.getOrDefault(node.id(), 0.0);
+                rank += (lastLayerRanks.get(parent) * this.alpha / size);
```

Comment:
should be degree instead of size.

degree is (sameLayerNodesV.size() + prevLayerNodesV.size() + adjacenciesV.size())